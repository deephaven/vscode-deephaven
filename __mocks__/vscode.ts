/* eslint-disable @typescript-eslint/naming-convention */
/**
 * Mock `vscode` module. Note that `vi.mock('vscode')` has to be explicitly
 * called in any test module needing to use this mock. It will not be loaded
 * automatically. This module also needs to be located under the `__mocks__`
 * directory under the project root to be found when `vi.mock('vscode')` is
 * called.
 */
import { vi } from 'vitest';

export class EventEmitter {
  fire = vi.fn().mockName('fire');
}

export class Position {
  constructor(line: number, character: number) {
    this.line = line;
    this.character = character;
  }

  readonly line: number;
  readonly character: number;
}

export enum QuickPickItemKind {
  Separator = -1,
  Default = 0,
}

export class Range {
  constructor(start: Position, end: Position);
  constructor(
    startLine: number,
    startCharacter: number,
    endLine: number,
    endCharacter: number
  );
  constructor(
    ...args: [Position, Position] | [number, number, number, number]
  ) {
    if (args.length === 2) {
      this.start = args[0];
      this.end = args[1];
    } else {
      const [startLine, startCharacter, endLine, endCharacter] = args;
      this.start = new Position(startLine, startCharacter);
      this.end = new Position(endLine, endCharacter);
    }
  }

  readonly start: Position;
  readonly end: Position;
}

export const ThemeColor = vi
  .fn()
  .mockName('ThemeColor')
  .mockImplementation((id: string) => ({
    id,
  }));

export const ThemeIcon = vi
  .fn()
  .mockName('ThemeIcon')
  .mockImplementation((id: string, color?: typeof ThemeColor) => ({
    id,
    color,
  }));

export enum TreeItemCollapsibleState {
  None = 0,
  Collapsed = 1,
  Expanded = 2,
}

export const window = {};

export const workspace = {
  getConfiguration: vi
    .fn()
    .mockName('getConfiguration')
    .mockReturnValue(new Map()),
  onDidChangeTextDocument: vi.fn().mockName('onDidChangeTextDocument'),
  onDidCloseTextDocument: vi.fn().mockName('onDidCloseTextDocument'),
};
