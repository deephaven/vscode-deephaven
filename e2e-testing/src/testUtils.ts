import {
  By,
  InputBox,
  TextEditor,
  TitleBar,
  until,
  VSBrowser,
  WebElement,
  WebView,
  type CodeLens,
  type Locator,
} from 'vscode-extension-tester';
import os from 'node:os';

export interface TabData {
  title: string;
  isSelected: boolean;
  isWebView: boolean;
}

export interface EditorGroupData {
  groupIndex: number;
  tabs: TabData[];
}

export interface WebViewExtended extends WebView {
  switchToContentFrame: (timeout?: number) => Promise<void>;
}

export async function elementExists(locator: Locator): Promise<boolean> {
  const { driver } = VSBrowser.instance;
  return (await driver.findElements(locator)).length > 0;
}

export async function elementExistsEventually(
  locator: Locator
): Promise<boolean> {
  const { driver } = VSBrowser.instance;
  await driver.wait(until.elementLocated(locator));
  return true;
}

export async function elementIsLazyLoaded(locator: Locator): Promise<boolean> {
  const existsInitially = await elementExists(locator);
  if (existsInitially) {
    throw new Error('Element is already present before lazy loading');
  }

  return elementExistsEventually(locator);
}

/**
 * Open a list of files in VS Code.
 * @param filePaths Paths of files to open (requires at least one path)
 */
export async function openFileResources(
  ...filePaths: [string, ...string[]]
): Promise<void> {
  if (['win32', 'darwin'].includes(os.platform())) {
    await VSBrowser.instance.openResources(...filePaths);
    return;
  }

  // In CI environment, openResources doesn't work on Linux. This workaround does.
  // https://github.com/redhat-developer/vscode-extension-tester/issues/506#issuecomment-1271156702
  const titleBar = new TitleBar();
  for (const filePath of filePaths) {
    const item = await titleBar.getItem('File');
    const fileMenu = await item!.select();
    const openItem = await fileMenu.getItem('Open File...');
    await openItem!.select();
    const input = await InputBox.create();
    await input.setText(filePath);
    await input.confirm();
  }
}

/**
 * Get a code lens based on title, or zero based index
 * @param editor Editor to get code lens from
 * @param indexOrTitle Index or title of code lens to get
 * @returns CodeLens
 */
export async function getCodeLens(
  editor: TextEditor,
  indexOrTitle: number | string
): Promise<CodeLens | undefined> {
  const ariaLabel = await editor.getAttribute('aria-label');

  // The `TextEditor.getCodeLens` method provided by `vscode-extension-tester`
  // does not seem to wait for the anchor element to be available, so we need
  // to wait for it ourselves. Including the `aria-label` to narrow down which
  // editor we are looking at.
  await VSBrowser.instance.driver.wait(
    until.elementLocated(
      By.css(
        `[aria-label="${ariaLabel}"] .contentWidgets span.codelens-decoration a`
      )
    )
  );

  return editor.getCodeLens(indexOrTitle);
}

/**
 * Switch to a frame based on the identifiers provided. If multiple identifiers
 * are provided, the function will switch nested frames. Each identifier can be
 * a numeric index of the frame, a CSS selector, or the actual iframe WebElement.
 * @param identifiers Identifiers of the nested frames to switch to
 * @returns A Promise that resolves when switching is complete.
 */
export async function switchToFrame(
  ...identifiers: [
    number | string | WebElement,
    ...(number | string | WebElement)[],
  ]
): Promise<void> {
  const { driver } = VSBrowser.instance;

  let iframe: WebElement | undefined;

  // Find the iframe element based on the identifier
  const findFrame = async (
    iframeOrIdentifier: string | number | WebElement
  ): Promise<WebElement | undefined> => {
    if (iframeOrIdentifier instanceof WebElement) {
      return iframeOrIdentifier;
    }

    if (typeof iframeOrIdentifier === 'number') {
      const iframes = await driver.findElements({
        tagName: 'iframe',
      });

      return iframes.at(iframeOrIdentifier);
    }

    if (typeof iframeOrIdentifier === 'string') {
      // Using `findElements` since `findElement` throws an error if no elements
      // are found
      const [iframe] = await driver.findElements(By.css(iframeOrIdentifier));
      return iframe;
    }
  };

  while (identifiers.length > 0) {
    const iframeOrIdentifier = identifiers.shift()!;

    iframe = await driver.wait(async () => findFrame(iframeOrIdentifier));

    if (iframe == null) {
      throw new Error(`Invalid iframe identifier ${iframeOrIdentifier}.`);
    }

    await driver.switchTo().frame(iframe);
  }
}
