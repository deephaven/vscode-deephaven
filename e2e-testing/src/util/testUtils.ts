import {
  By,
  InputBox,
  SideBarView,
  TitleBar,
  until,
  VSBrowser,
  WebElement,
  type CodeLens,
  type Locator,
  type TextEditor,
  type ViewItem,
} from 'vscode-extension-tester';
import os from 'node:os';
import { RETRY_SWITCH_IFRAME_ERRORS } from './constants';

export interface TabData {
  title: string;
  isSelected: boolean;
  isWebView: boolean;
}

export interface WebViewData {
  isVisible?: true;
  hasContent?: true;
}

export interface EditorGroupData {
  groupIndex: number;
  tabs: TabData[];
  webViews?: WebViewData[];
}

/**
 * Selector for an iframe. Can be a numeric index in the frames collection of
 * current context, a CSS selector string, or a function that returns a Promise
 * resolving to a WebElement.
 */
export type FrameSelector = number | string | (() => Promise<WebElement>);

/**
 * Disconnect from Deephaven server by clicking on disconnect action on server
 * node.
 * @param title Title of server node to disconnect from
 */
export async function disconnectFromServer(title: string): Promise<void> {
  const serverViewItem = await getSidebarViewItem('Servers', title);
  await serverViewItem?.select();
  const disconnectAction = await serverViewItem?.findElement(
    By.css('[aria-label="Disconnect from Server"]')
  );
  await disconnectAction?.click();
}

/**
 * Check if an element exists in the current context. This is needed since the
 * default Selenium driver `findElement` methods throw if element not found.
 * @param locator Locator of element to check
 * @returns True if element exists, false otherwise
 */
export async function elementExists(locator: Locator): Promise<boolean> {
  const { driver } = VSBrowser.instance;
  return (await driver.findElements(locator)).length > 0;
}

/**
 * We don't have access to some of the underlying error classes from
 * vscode-extension-tester, but we can approximate the type of error by grabbing
 * any text before the first colon.
 * @param error Error to extract type from
 * @returns Error type
 */
export function extractErrorType(error: unknown): string {
  return String(error).split(':')[0];
}

/**
 * Get a sidebar view item based on section and title.
 * @param section Section of sidebar view to get item from
 * @param title Title of item to get
 * @returns Sidebar view item
 */
export async function getSidebarViewItem(
  section: string,
  title: string
): Promise<ViewItem | undefined> {
  const sideBarView = new SideBarView();
  const serverSection = await sideBarView.getContent().getSection(section);
  return serverSection.findItem(title);
}

/**
 * Get a code lens based on title, or zero based index
 * @param editor Editor to get code lens from
 * @param indexOrTitle Index or title of code lens to get
 * @returns CodeLens
 */
export async function getCodeLens(
  editor: TextEditor,
  indexOrTitle: number | string
): Promise<CodeLens> {
  // The `TextEditor.getCodeLens` method provided by `vscode-extension-tester`
  // does not seem to explicitly wait for the element to be available, which
  // sometimes works, and sometimes does not. To be safe, we need wait for it
  // ourselves.
  return VSBrowser.instance.driver.wait<CodeLens>(async () =>
    editor.getCodeLens(indexOrTitle)
  );
}

/**
 * Open a list of files in VS Code.
 * @param filePaths Paths of files to open (requires at least one path)
 */
export async function openFileResources(
  ...filePaths: [string, ...string[]]
): Promise<void> {
  if (['win32', 'darwin'].includes(os.platform())) {
    await VSBrowser.instance.openResources(...filePaths);
    return;
  }

  // In CI environment, openResources doesn't work on Linux. This workaround does.
  // https://github.com/redhat-developer/vscode-extension-tester/issues/506#issuecomment-1271156702
  const titleBar = new TitleBar();
  for (const filePath of filePaths) {
    const item = await titleBar.getItem('File');
    const fileMenu = await item!.select();
    const openItem = await fileMenu.getItem('Open File...');
    await openItem!.select();
    const input = await InputBox.create();
    await input.setText(filePath);
    await input.confirm();
  }
}

/**
 * Execute "Run Deephaven File" code lens in the given editor.
 * @param editor Editor to execute code lens in
 */
export async function runDhFileCodeLens(editor: TextEditor): Promise<void> {
  const runDhFileCodeLens = await getCodeLens(editor, 'Run Deephaven File');
  await runDhFileCodeLens.click();
}

/**
 * Wrapper a labeled step in sequential code.
 * @param n step number
 * @param label step label
 * @param fn code to execute
 */
export async function step<TResult>(
  n: number,
  label: string,
  fn: (stepLabel: string) => Promise<TResult>
): Promise<TResult> {
  const stepLabel = `Step ${n}: ${label}`;
  // eslint-disable-next-line no-console
  console.log(stepLabel);
  return fn(stepLabel);
}

/**
 * Switch to a frame based on the identifiers provided. If multiple identifiers
 * are provided, the function will switch nested frames. Each identifier can be
 * a numeric index of the frame, a CSS selector, or the actual iframe WebElement.
 * @param identifiers Identifiers of the nested frames to switch to
 * @returns A Promise that resolves when switching is complete.
 */
export async function switchToFrame(
  identifiers: [FrameSelector, ...FrameSelector[]],
  timeout?: number
): Promise<void> {
  const { driver } = VSBrowser.instance;

  let iframe: WebElement | undefined;

  // Find the iframe element based on the identifier
  const findFrame = async (
    iframeOrIdentifier: FrameSelector
  ): Promise<WebElement | undefined> => {
    if (typeof iframeOrIdentifier === 'number') {
      const iframes = await driver.findElements({
        tagName: 'iframe',
      });

      return iframes.at(iframeOrIdentifier);
    }

    if (typeof iframeOrIdentifier === 'string') {
      // Using `findElements` since `findElement` throws an error if no elements
      // are found
      const [iframe] = await driver.findElements(By.css(iframeOrIdentifier));
      return iframe;
    }

    return iframeOrIdentifier();
  };

  while (identifiers.length > 0) {
    const iframeOrIdentifier = identifiers.shift()!;

    iframe = await driver.wait<WebElement>(
      async () => findFrame(iframeOrIdentifier),
      timeout
    );

    try {
      await driver.switchTo().frame(iframe);
    } catch (err) {
      const errorType = extractErrorType(err);

      if (!RETRY_SWITCH_IFRAME_ERRORS.has(errorType)) {
        throw err;
      }

      // eslint-disable-next-line no-console
      console.log(`Retrying after '${errorType}' error`);

      // Try retrieving the iframe and switching again
      iframe = await driver.wait<WebElement>(
        async () => findFrame(iframeOrIdentifier),
        timeout
      );

      try {
        await driver.switchTo().frame(iframe);
      } catch (err) {
        // eslint-disable-next-line no-console
        console.log(`Failed to switch to frame: ${iframeOrIdentifier}`, err);
        throw err;
      }
    }
  }
}
