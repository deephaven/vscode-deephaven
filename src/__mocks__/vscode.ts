/**
 * Mock `vscode` module. Note that `vi.mock('vscode')` has to be explicitly
 * called in any test module needing to use this mock. It will not be loaded
 * automatically.
 */
import { vi } from 'vitest';

export const workspace = {
  getConfiguration: vi.fn().mockReturnValue(new Map()),
};
