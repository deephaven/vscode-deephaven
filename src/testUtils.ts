/**
 * This module contains test utils that should only be consumed by `.spec.ts`
 * files, but since it's not a `.spec.ts` file, it is special cased to exclude
 * from the extension build (tsconfig.json) and include in the unit test build
 * (tsconfig.unit.json). It shouldn't be imported by the extension code or have
 * any dependencies on the extension code.
 */

import { expect, vi } from 'vitest';
import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

export type MessageHandler<TData> = (
  event: Partial<MessageEvent<TData>>
) => unknown;

/**
 * Return last registered event handler for 'message' event. Assumes that
 * `window.addEventListener` has been mocked or spied on.
 */
export function getLastMessageHandler<TData>(): MessageHandler<TData> {
  const messageCalls = vi
    .mocked(window.addEventListener<'message'>)
    .mock.calls.filter(([type]) => type === 'message');

  expect(messageCalls.length).toBeGreaterThan(0);

  return messageCalls.at(-1)![1] as MessageHandler<TData>;
}

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
