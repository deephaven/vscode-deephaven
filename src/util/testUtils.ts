import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
