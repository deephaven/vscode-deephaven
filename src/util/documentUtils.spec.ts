import * as vscode from 'vscode';
import { describe, it, expect, vi } from 'vitest';
import {
  deserializeRange,
  normalizeLanguageId,
  parseMarkdownCodeblocks,
  serializeRange,
} from './documentUtils';
import { mockDocument } from './testUtils';
import type { SerializedRange } from '../types';

vi.mock('vscode');

const markdownDocContent = [
  '```py',
  'from deephaven import time_table',
  '',
  'simple_ticking = time_table("PT2S")',
  '```',
  '',
  '```python',
  'from deephaven import time_table',
  '',
  'simple_ticking = time_table("PT2S")',
  '```',
  '',
  '## Groovy Example',
  '',
  '```groovy',
  'simple_ticking = timeTable("PT2S")',
  '```',
].join('\n');

describe('parseMarkdownCodeblocks', () => {
  it('should parse Deephaven code blocks from a Markdown document', () => {
    const codeBlocks = parseMarkdownCodeblocks(
      mockDocument('test.md', 1, markdownDocContent)
    );

    expect(codeBlocks).toMatchSnapshot();
  });
});

describe('normalizeLanguageId', () => {
  it.each([
    ['python', 'python'],
    ['py', 'python'],
    ['groovy', 'groovy'],
    ['other', 'other'],
  ])('should normalize languageId: %s->%s', (given, expected) => {
    const normalized = normalizeLanguageId(given);
    expect(normalized).toBe(expected);
  });
});

describe('deserializeRange', () => {
  it('should deserialize a SerializedRange object to a vscode.Range', () => {
    const serializedRange: SerializedRange = [
      { line: 1, character: 4 },
      { line: 3, character: 9 },
    ];

    const actual = deserializeRange(serializedRange);

    expect(actual).toEqual(
      new vscode.Range(
        serializedRange[0].line,
        serializedRange[0].character,
        serializedRange[1].line,
        serializedRange[1].character
      )
    );
  });
});

describe('serializeRange', () => {
  it('should serialize a vscode.Range object to a SerializedRange', () => {
    const range = new vscode.Range(1, 4, 3, 9);
    const actual = serializeRange(range);

    expect(actual).toEqual([
      { line: range.start.line, character: range.start.character },
      { line: range.end.line, character: range.end.character },
    ]);
  });
});
