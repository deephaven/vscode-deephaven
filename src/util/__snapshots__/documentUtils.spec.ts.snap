// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseMarkdownCodeblocks > should parse Deephaven code blocks from a Markdown document 1`] = `
[
  {
    "languageId": "python",
    "range": Range {
      "end": Position {
        "character": 35,
        "line": 3,
      },
      "start": Position {
        "character": 0,
        "line": 1,
      },
    },
  },
  {
    "languageId": "python",
    "range": Range {
      "end": Position {
        "character": 35,
        "line": 9,
      },
      "start": Position {
        "character": 0,
        "line": 7,
      },
    },
  },
  {
    "languageId": "groovy",
    "range": Range {
      "end": Position {
        "character": 34,
        "line": 15,
      },
      "start": Position {
        "character": 0,
        "line": 15,
      },
    },
  },
]
`;
